### DISCUSSION TITLE
rendering quarto in rstudio

### DISCUSSION THREAD

**Discussion by olechnwin** [quarto-dev/quarto-cli#6924] [Q&A]
### Description

When rendering a quarto document, is there a way to make all the variables defined in quarto be available in my current R session?
That way it makes it easier to continue developing the script/document.

For example (see image below) how do I make x available after rendering test.qmd?

![image](https://github.com/quarto-dev/quarto-cli/assets/20069067/daee51d5-de4e-4cd1-8d91-37d2e918857a)

**Comment by mcanouil** [comment]
Execute/run the cells as you go using the UI: the icons at the right of the code cells and the "Run" button in the bar below the files tabs.
(Same thing as with rmarkdown documents)

Note that Quarto does not run inside R but outside contrary to rmarkdown, thus the R session is completely independent.

**Reply by olechnwin** [reply]
I'm hoping for another way to do this. If I render entire quarto document which makes it run the entire code, I'll have to wait for it to finish executing everything. To then have to re-run all the cells using the UI seems like doing a redundant thing that takes a long time. Also, for long computation, this is very time consuming.

**Reply by olechnwin** [reply]
@mcanouil, you mentioned that Quarto doesn't run inside R. Does that mean if I switched back to rmarkdown, I can have the objects available in my current R session after rendering?

**Reply by mcanouil** [reply]
For short, no you can't do `quarto::quarto_render()` and get as a side effect the objects (yes, it's a side effect not a feature from `rmarkdown::render()`).

If you are interactively developing, why are you rendering the document? I don't get it.

I suspect you are using a side effect or `rmarkdown::render` in the R terminal which leave the object there.
You can't do that with Quarto.
To note, even if it seems "convenient", this basically break any "guarantee" of reproducibility as you also use anything in the current environment.
In my opinion, it is not a good practice.

The proper way to go is to execute cells one by one or all of them in one go without rendering the document.
For long computations there are various ways to deal with them, use `targets` for example.

Also I don't understand what difference it makes to render the rmarkdown document inside the R session compared to using the run all cells. The computations are the same and will take the exact same amount of time.

**Reply by olechnwin** [reply]
I am rendering the document to create kind of like a progress report. My project usually runs for months but I'll have to create a progress report weekly. I actually just realized `rmarkdown::render` has this side effect.

I do understand your point about it not a good practice since it breaks reproducibility. Is there a good practice guide on this for rendering and adding new code ? I'm not familiar with `targets`. Where can I find more info on this?

"Also I don't understand what difference it makes to render the rmarkdown document inside the R session compared to using the run all cells. The computations are the same and will take the exact same amount of time." I just meant if I can have all the objects available after rendering the document then I don't have to re-run the same code again to make the objects available in my R session.

**Reply by mcanouil** [reply]
In your case, I strongly suggest to use `targets` (or anything similar) which is compatible with both rmarkdown/Quarto.
With this package, you "never" execute again something up to date.
See <https://books.ropensci.org/targets/>.

Also note that "rendering" is not "developing", the two steps should really be independent.

**Reply by olechnwin** [reply]
Thanks for the link to `targets`.
Sounds like what you are suggesting is not to render until I'm done with the code. Is there another way to generate a one page report, then?

**Reply by mcanouil** [reply]
Developing -> run each cells.
Rendering -> generate the report.
